Silicon/CommonSocPkg/Library/WatchDogTimerLib/WatchDogTimerLib.c

```diff
@@ -238,3 +258,55 @@ IsWdtLocked (
   }
 }
 
+/**
+  Get the number of failed boots.
+
+  @retval UINT32              the number of boots
+
+**/
+UINT32
+EFIAPI
+GetFailedBootCount (
+  VOID
+  )
+{
+  return WdtGetScratchpad (BOOT_COUNT_MASK) >> 18;
+}
+
+/**
+  Increment the number of failed boots.
+**/
+VOID
+EFIAPI
+IncrementFailedBootCount (
+  VOID
+  )
+{
+  UINT32 ReadBack;
+
+  ReadBack = WdtGetScratchpad (BOOT_COUNT_MASK);
+
+  // Increment counter
+  ReadBack += (1 << 18);
+
+  // Set bits
+  WdtSetScratchpad ((ReadBack & BOOT_COUNT_MASK));
+
+  // Invert counter
+  ReadBack = ~ReadBack;
+
+  // Clear bits
+  WdtClearScratchpad ((ReadBack & BOOT_COUNT_MASK));
```

Comment:
Because the existing WdtSetScratchpad() exclusively sets bits -- it doesn't clear them. WdtClearScratchpad() is needed for this. Say we want to increment the counter from 0b01 to 0b10. If we call WdtSetScratchpad(0b10 << B_ACPI_IO_OC_WDT_CTL_SCRATCHPAD_BOOT_CNT_SHIFT) when the counter is at 0b01, the result will be the counter at 0b11. We need to call WdtClearScratchpad(0b1 << B_ACPI_IO_OC_WDT_CTL_SCRATCHPAD_BOOT_CNT_SHIFT) to fix this. I adjusted this API to use GetFailedBootCount() and ClearFailedBootCount() to make things a bit cleaner.